juju/juju
# [JUJU-299] Store unit CharmURL as a string reference

We have long stored in the unit Mongo document, an actual `*charm.URL` reference.

This of course is fraught with danger, as upstream changes to the library can change what is stored, and potentially render old documents incompatible with an updated type.

In addition, we instituted `GetBSON` and `SetBSON` helpers to handle persistence and retrieval, which were made non-trivial by the introduction of CharmHub and multiple possible formats for this URL.

We have ended up in a situation where we are paying a small overhead to parse this URL upon every access, even when we might not use the field.

Here we update the document to store a string, and only parse the URL at need. The patch is scoped only to what is necessary to change the one field in the unit document. We still do unnecessary conversion, and could benefit from changing the application document as well, including all the sundry methods that accept `*charm.URL`, but immediately stringify it for use.

Upgrades had one usage of the doc for a version 2.8 step. This is replaced with the old version of the doc still storing a `*charm.URL`.

## QA steps

- Bootstrap a 2.7 controller and `juju install ubuntu -n 3`.
- With this patch, `juju upgrade-controller --build-agent`.
- Run `juju upgrade-model` twice. It will jump to the latest 2.8, then this patch.
- Ensure that the logs show no errors relating to unit documents.

## Documentation changes

None.

## Bug reference

N/A
